### ISSUE TITLE
anomaly when invoking Guarded with shelved subproof.

### ISSUE THREAD

**Issue by PierreCorbineau** [coq/coq#19634]
### Description of the problem

When defining a fixpoint interactively (with tactics), checking guardedness with shelved subgoals triggers an anomaly.

Error:
Anomaly
"File "kernel/cClosure.ml", line 134, characters 26-32: Assertion failed."

### Small Coq file to reproduce the bug

```coq
Fixpoint f (n:nat) {struct n}: nat.
Proof.
assert(nn:nat) by shelve.
pose (fnn:=f nn).

Guarded. (* anomaly *)
```

### Version of Coq where this bug occurs

8.19.1

**Comment by SkySkimmer** [comment]
Seems fine on master

**Comment by silene** [comment]
It also works fine on 8.20. So, it seems to have been fixed between 8.19 and 8.20.
